paasta_tools/setup_istio_mesh.py

```diff
@@ -0,0 +1,245 @@
+)
+
+
+class ErrorCreatingUnifiedService(Exception):
+    pass
+
+
+class ErrorGettingServiceList(Exception):
+    pass
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(description="Creates Kubernetes services.")
+    parser.add_argument(
+        "-v", "--verbose", action="store_true", dest="verbose", default=False,
+    )
+    parser.add_argument(
+        "--dry-run", action="store_true", dest="dry_run", default=False,
+    )
+    parser.add_argument(
+        "-l",
+        "--rate-limit",
+        dest="rate_limit",
+        default=0,
+        metavar="LIMIT",
+        type=int,
+        help="Update or create up to this number of service instances. Default is 0 (no limit).",
+    )
+    parser.add_argument(
+        "-f",
+        "--file-path",
+        dest="file_path",
+        default="/nail/etc/services/services.json",
+        metavar="LIMIT",
+        type=str,
+        help="services file to use for smartstack namespace. Default file is /nail/etc/services/services.json",
+    )
+    args = parser.parse_args()
+    return args
+
+
+def build_port_namespace_mapping(file_path: str) -> Mapping:
+    paasta_namespaces = {}
+    with open(file_path) as service_file:
+        smartstack_namespaces = json.load(service_file)
+        for namespace, endpoint in smartstack_namespaces.items():
+            paasta_namespaces[endpoint["port"]] = namespace
+    return paasta_namespaces
+
+
+def sanitise_kubernetes_service_name(name: str) -> str:
+    name = sanitise_kubernetes_name(name)
+    return name.replace(".", "--")
+
+
+def get_existing_kubernetes_service_names(kube_client: KubeClient) -> Set[str]:
+    label_selector = "paasta.yelp.com/owner=istio"
+    service_objects = kube_client.core.list_namespaced_service(
+        "paasta", label_selector=label_selector
+    )
+    if not service_objects:
+        raise ErrorGettingServiceList("Error retrieving services list from k8s api")
+
+    return {item.metadata.name for item in service_objects.items}
+
+
+def setup_unified_service(
+    kube_client: KubeClient, port_list: AbstractSet
+) -> k8s.V1Service:
+
+    # Add port 1337 for envoy unified listener.
+    # Clients can connect to this listenner and set x-yelp-svc header for routing
+    ports = [
+        k8s.V1ServicePort(
+            name=f"p{UNIFIED_SVC_PORT}",
+            port=UNIFIED_SVC_PORT,
+            protocol="TCP",
+            target_port=PAASTA_SVC_PORT,
+            app_protocol="http",
+        )
+    ]
+
+    # Add smartstack ports for routing, Clients can connect to this
+    # Directly without need of setting x-yelp-svc header
+    for port in port_list:
+        port_spec = k8s.V1ServicePort(
+            name=f"p{port}",
+            port=port,
+            protocol="TCP",
+            target_port=PAASTA_SVC_PORT,
+            app_protocol="http",
+        )
+        ports.append(port_spec)
+
+    service_labels: KubeSvcLabels = {
+        "paasta.yelp.com/owner": "istio",
```

Comment:
i assume this attribute will be used to garbage-collect services, but `owner` is too generic and vague for this purpose, let's use something specific like `managed_by=setup_istio_mesh.py` and put it in annotations to not pollute labels (we don't really need to index on that label as we're assuming that every service in paasta namespace is managed by this script)